vmdb/app/models/miq_ae_class.rb

```diff
@@ -153,8 +153,28 @@ def self.copy(options)
     end
   end
 
+  def self.waypoint_ids_for_state_machines
+    ids = []
+    MiqAeClass.all.select(&:state_machine?).each do |klass|
+      ids << klass.id
+      next if ids.include?(klass.ae_namespace.id)
+      sub_namespaces(klass.ae_namespace, ids)
+    end
+    ids
+  end
+
   private
 
+  def self.sub_namespaces(ns_obj, ids)
+    loop do
+      break if ns_obj.nil? || (ns_obj && ids.include?(ns_obj.id))
```

Comment:
`break if ns_obj.nil? || ids.include?(ns_obj.id)`
is enough due to lazy evaluation